File: CodeGenDotNet 159059-1372 (a)/CSharp/Component Shells/WinSupport/BaseEditUserControl.cs
// ====================================================================
//  Copyright ©2004, Kathleen Dollard, All Rights Reserved.
// ====================================================================
//   I'm distributing this code so you'll be able to use it to see code
//   generation in action and I hope it will be useful and you'll enjoy 
//   using it. This code is provided "AS IS" without warranty, either 
//   expressed or implied, including implied warranties of merchantability 
//   and/or fitness for a particular purpose. 
//  ====================================================================
//  Summary: Base class for WinForms editing user controls


using System;

namespace KADGen.WinSupport
{
	public class BaseEditUserControl : System.Windows.Forms.UserControl
	{

		// Implements IEditUserControl

		public delegate void DataChangedHandler(
			object sender,
			EventArgs e );
		public event DataChangedHandler DataChanged;

		#region Class Level Declaration
		private BaseSelectUserControl mSelectUserControl;
		protected BusinessObjectSupport.IBusinessObject mBusinessObject;
		protected int mMargin = 5;
		protected bool mbIsDirty;
		protected string mCaption;
		// protected mMinimumSize System.Drawing.Size
		#endregion

		#region  Windows Form Designer generated code 

		public BaseEditUserControl() : base()
		{

			//This call is required by the Windows Form Designer.
			InitializeComponent();

			//Add any initialization after the InitializeComponent() call

		}

		//Form overrides dispose to clean up the component list.
		protected override void Dispose( bool disposing )
		{
			if( disposing )
			{
				if( components != null) 
				{
					components.Dispose();
				}
			}
			base.Dispose(disposing);
		}

		//Required by the Windows Form Designer
		private System.ComponentModel.IContainer components;

		//NOTE: The following procedure is required by the Windows Form Designer
		//It can be modified using the Windows Form Designer.  
		//Do not modify it using the code editor.
		protected internal System.Windows.Forms.ErrorProvider errProvider;
		protected internal System.Windows.Forms.ToolTip toolTip;
		[System.Diagnostics.DebuggerStepThrough()] private void InitializeComponent()
		{
			this.errProvider = new System.Windows.Forms.ErrorProvider();
			this.toolTip = new System.Windows.Forms.ToolTip();
			//
			//errProvider
			//
			this.errProvider.ContainerControl = this;
			//
			//toolTip
			//
			//
			//BaseEditUserControl
			//
			this.Name = "BaseEditUserControl";

		}

		#endregion

		#region Public Properties and Methods
		public BaseSelectUserControl SelectUserControl
		{
			get
			{
				return mSelectUserControl;
			}
		}

		public virtual CSLA.ReadOnlyCollectionBase GetList() 
		{
			throw new System.ApplicationException("The GetList method must be overridden");
		}

		public void VisitControls()
		{
			VisitControls(this);
		}
		public void VisitControls( System.Windows.Forms.Control ctl )
		{
			//Visit every control on the form to set error provider
			//By doing this as a complete first step , all error flags are set
			foreach( System.Windows.Forms.Control childCtl in ctl.Controls )
			{
				childCtl.Focus();
				if( childCtl.Controls.Count > 0 )
				{
					VisitControls(childCtl);
				}
			}
		}

		public bool IsFormValid() 
		{
			return IsFormValid(this , this.errProvider);
		}
		public bool IsFormValid(
			System.Windows.Forms.Control ctl ,
			System.Windows.Forms.ErrorProvider errProv ) 
		{
			bool bValid = true;
			foreach( System.Windows.Forms.Control childCtl in ctl.Controls )
			{
				//childCtrl.Focus()
				if( errProv.GetError(childCtl) != "" )
				{
					bValid = false;
					break;
				}
				if( childCtl.Controls.Count > 0 )
				{
					bValid = IsFormValid(childCtl , errProv);
					if( ! bValid )
					{
						break;
					}
				}
			}
			return bValid;
		}

		#endregion

		#region Protected Properties and Methods
		protected virtual void ResizeUC()
		{
			int maxHeight = 0;
			foreach( System.Windows.Forms.Control ctl in this.Controls )
			{
				if( ctl.Visible )
				{
					if( ctl.Bottom > maxHeight )
					{
						maxHeight = ctl.Bottom;
					}
				}
			}
			this.ClientSize = new System.Drawing.Size(this.ClientSize.Width , maxHeight + mMargin);
		}

		protected virtual void OnDataChanged( object sender, EventArgs e )
		{
			DataChanged(sender , e);
		}

		protected void RemoveControl(
			System.Windows.Forms.Control ctl,
			string sFieldName )
		{
			int height = 0;
			if( sFieldName.Trim().Length > 0 )
			{
				if( ctl.Name.Length > 3 )
				{
					if( ctl.Name.Substring(3) == sFieldName )
					{
						// remove later controls and exit               
						ctl.Visible = false;
						if( ctl.Name.Substring(0 , 3).ToLower() != "lbl" )
						{
							height = ctl.Height + ((IEditUserControl) this).VerticalMargin;
							foreach( System.Windows.Forms.Control ctlOther in ctl.Parent.Controls )
							{
								if( ctlOther.Top > ctl.Bottom )
								{
									ctlOther.Top -= height;
								}
							}
						}
					}
				}
				foreach( System.Windows.Forms.Control ctlChild in ctl.Controls )
				{
					RemoveControl(ctlChild , sFieldName);
				}
			}
		}

		protected virtual void BindField(
			System.Windows.Forms.Control control ,
			string propertyName ,
			object dataSource ,
			string  dataMember,
			System.Type netType )
		{

			System.Windows.Forms.Binding bd;

			for( int index = control.DataBindings.Count - 1; index >=0 ; index-- )
			{
				bd = control.DataBindings[index];
				if( bd.PropertyName == propertyName )
				{
					control.DataBindings.Remove(bd);
				}
			}
			control.DataBindings.Add(propertyName , dataSource , dataMember);

			BindEvents(control , propertyName , dataSource , dataMember , netType);

		}

		protected virtual void BindEvents(
			System.Windows.Forms.Control control,
			string controlProperty ,
			object source ,
			string sourceProperty ,
			System.Type netType )
		{
			if( netType == typeof(System.Decimal) )
			{
				WinSupport.Utility.BindEvents(control , controlProperty , source , sourceProperty ,
					new System.Windows.Forms.ConvertEventHandler(FormatCurrency) , 
					new System.Windows.Forms.ConvertEventHandler(ParseCurrency));
			}
		}


		#region Format Event Handlers
		protected virtual void FormatCurrency(
			object sender ,
			System.Windows.Forms.ConvertEventArgs e )
		{
			e.Value = ((System.Decimal) e.Value).ToString("C");
		}

		protected virtual void ParseCurrency(
			object sender ,
			System.Windows.Forms.ConvertEventArgs e )
		{
			string val = e.Value.ToString();
			val = val.Replace("$" , "");
			e.Value = System.Decimal.Parse(val);
		}

		#endregion

		public string Caption
		{
			get
			{
				return mCaption;
			}
		}

		#endregion

	}
}

File: CodeGenDotNet 159059-1372 (a)/VBCode/Component Shells/WinSupport/BaseEditUserControl.vb
' ====================================================================
'  Copyright ©2004, Kathleen Dollard, All Rights Reserved.
' ====================================================================
'   I'm distributing this code so you'll be able to use it to see code
'   generation in action and I hope it will be useful and you'll enjoy 
'   using it. This code is provided "AS IS" without warranty, either 
'   expressed or implied, including implied warranties of merchantability 
'   and/or fitness for a particular purpose. 
'  ====================================================================
'  Summary: Base class for WinForms editing user controls

Option Strict On
Option Explicit On 

Imports System

Public Class BaseEditUserControl
   Inherits System.Windows.Forms.UserControl
   ' Implements IEditUserControl

   Public Event DataChanged( _
                  ByVal sender As Object, _
                  ByVal e As EventArgs)


#Region "Class Level Declaration"
   Private mSelectUserControl As BaseSelectUserControl
   Protected mBusinessObject As BusinessObjectSupport.IBusinessObject
   Protected mMargin As Int32 = 5
   Protected mbIsDirty As Boolean
   Protected mCaption As String
   ' Protected mMinimumSize As System.Drawing.Size
#End Region

#Region " Windows Form Designer generated code "

   Public Sub New()
      MyBase.New()

      'This call is required by the Windows Form Designer.
      InitializeComponent()

      'Add any initialization after the InitializeComponent() call

   End Sub

   'Form overrides dispose to clean up the component list.
   Protected Overloads Overrides Sub Dispose(ByVal disposing As Boolean)
      If disposing Then
         If Not (components Is Nothing) Then
            components.Dispose()
         End If
      End If
      MyBase.Dispose(disposing)
   End Sub

   'Required by the Windows Form Designer
   Private components As System.ComponentModel.IContainer

   'NOTE: The following procedure is required by the Windows Form Designer
   'It can be modified using the Windows Form Designer.  
   'Do not modify it using the code editor.
   Protected Friend WithEvents errProvider As System.Windows.Forms.ErrorProvider
   Protected Friend WithEvents toolTip As System.Windows.Forms.ToolTip
   <System.Diagnostics.DebuggerStepThrough()> Private Sub InitializeComponent()
      Me.errProvider = New System.Windows.Forms.ErrorProvider
      Me.toolTip = New System.Windows.Forms.ToolTip
      '
      'errProvider
      '
      Me.errProvider.ContainerControl = Me
      '
      'toolTip
      '
      '
      'BaseEditUserControl
      '
      Me.Name = "BaseEditUserControl"

   End Sub

#End Region

#Region "Public Properties and Methods"
   Public ReadOnly Property SelectUserControl() As BaseSelectUserControl
      Get
         Return mSelectUserControl
      End Get
   End Property

   Public Overridable Function GetList() As CSLA.ReadOnlyCollectionBase
      Throw New System.ApplicationException("The GetList method must be overridden")
   End Function

   Public Sub VisitControls()
      VisitControls(Me)
   End Sub
   Public Sub VisitControls(ByVal ctl As Windows.Forms.Control)
      'Visit every control on the form to set error provider
      'By doing this as a complete first step, all error flags are set
      Dim childCtl As Windows.Forms.Control
      For Each childCtl In ctl.Controls
         childCtl.Focus()
         If childCtl.Controls.Count > 0 Then
            VisitControls(childCtl)
         End If
      Next
   End Sub

   Public Function IsFormValid() As Boolean
      Return IsFormValid(Me, Me.errProvider)
   End Function
   Public Function IsFormValid( _
               ByVal ctl As Windows.Forms.Control, _
               ByVal errProv As Windows.Forms.ErrorProvider) As Boolean
      Dim bValid As Boolean = True
      Dim childCtl As Windows.Forms.Control
      For Each childCtl In ctl.Controls
         'childCtrl.Focus()
         If errProv.GetError(childCtl) <> "" Then
            bValid = False
            Exit For
         End If
         If childCtl.Controls.Count > 0 Then
            bValid = IsFormValid(childCtl, errProv)
            If Not bValid Then
               Exit For
            End If
         End If
      Next
      Return bValid
   End Function

#End Region

#Region "Protected Properties and Methods"
   Protected Overridable Sub ResizeUC()
      Dim maxHeight As Integer
      For Each ctl As Windows.Forms.Control In Me.Controls
         If ctl.Visible Then
            If ctl.Bottom > maxHeight Then
               maxHeight = ctl.Bottom
            End If
         End If
      Next
      Me.ClientSize = New Drawing.Size(Me.ClientSize.Width, maxHeight + mMargin)
   End Sub

   Protected Overridable Sub OnDataChanged(ByVal sender As Object, ByVal e As EventArgs)
      RaiseEvent DataChanged(sender, e)
   End Sub

   Protected Sub RemoveControl( _
               ByVal ctl As Windows.Forms.Control, _
               ByVal sFieldName As String)
      Dim height As Int32
      If sFieldName.Trim.Length > 0 Then
         If ctl.Name.Length > 3 Then
            If ctl.Name.Substring(3) = sFieldName Then
               ' remove later controls and exit
               ctl.Visible = False
               If ctl.Name.Substring(0, 3).ToLower <> "lbl" Then
                  height = ctl.Height + CType(Me, IEditUserControl).VerticalMargin
                  For Each ctlOther As Windows.Forms.Control In ctl.Parent.Controls
                     If ctlOther.Top > ctl.Bottom Then
                        ctlOther.Top -= height
                     End If
                  Next
               End If
            End If
         End If
         For Each ctlChild As Windows.Forms.Control In ctl.Controls
            RemoveControl(ctlChild, sFieldName)
         Next
      End If
   End Sub

   Protected Overridable Sub BindField( _
               ByVal control As Windows.forms.Control, _
               ByVal propertyName As String, _
               ByVal dataSource As Object, _
               ByVal dataMember As String, _
               ByVal netType As System.Type)

      Dim bd As Windows.Forms.Binding
      Dim index As Integer

      For index = control.DataBindings.Count - 1 To 0 Step -1
         bd = control.DataBindings.Item(index)
         If bd.PropertyName = propertyName Then
            control.DataBindings.Remove(bd)
         End If
      Next
      control.DataBindings.Add(propertyName, dataSource, dataMember)

      BindEvents(control, propertyName, dataSource, dataMember, netType)

   End Sub

   Protected Overridable Sub BindEvents( _
               ByVal control As Windows.forms.Control, _
               ByVal controlProperty As String, _
               ByVal source As Object, _
               ByVal sourceProperty As String, _
               ByVal netType As System.Type)
      If netType Is GetType(System.Decimal) Then
         WinSupport.Utility.BindEvents(control, controlProperty, source, sourceProperty, _
                  AddressOf FormatCurrency, AddressOf ParseCurrency)
      End If
   End Sub


#Region "Format Event Handlers"
   Protected Overridable Sub FormatCurrency( _
                  ByVal sender As Object, _
                  ByVal e As Windows.Forms.ConvertEventArgs)
      e.Value = CType(e.Value, System.Decimal).ToString("C")
   End Sub

   Protected Overridable Sub ParseCurrency( _
               ByVal sender As Object, _
               ByVal e As Windows.Forms.ConvertEventArgs)
      Dim val As String = e.Value.ToString
      val = val.Replace("$", "")
      e.Value = System.Decimal.Parse(val)
   End Sub

#End Region

   Public ReadOnly Property Caption() As String
      Get
         Return mCaption
      End Get
   End Property

#End Region

End Class
